Core: Fine grained recomputes

pieterhijma opened the pull request:
This PR introduces fine-grained recomputes in FreeCAD. These are recomputes based on dependencies between *properties* instead of based on *objects* as is the case in FreeCAD currently. This is essentially the [implementation of phase 1](https://github.com/pieterhijma/FreeCAD-Enhancement-Proposals/blob/variant-parts/FEPs/FEP-0010-variant-parts/README.md#phase-1-fine-grained-recomputation) of [FEP-0010](https://github.com/pieterhijma/FreeCAD-Enhancement-Proposals/blob/variant-parts/FEPs/FEP-0010-variant-parts/README.md).

Important design decisions:
- I chose to deprecate `Property::Status::Single`, a non-used flag that indicates that floating points in some properties need to be single precision instead of double precision. As far as I can tell, this is not actively used in FreeCAD. Removing this flag paves the way for introducing an `Property::Status::Input` flag that allows us to mark properties as "true" input properties.
- It introduces a simple dependency edge tuple with a "from object", "from property", "to object", "to property". This is used to define the fine-grained relations in the dependency graph.
- Besides the regular InList and OutList, FreeCAD now computes InListProp and OutListProp, the same datastructure but with more fine-grained information.
- There is a preference in the Gui that allows users to turn on or off fine-grained recomputes. From the console or Python use `FreeCAD.ParamGet("User parameter:BaseApp/Preferences/General").SetBool("FineGrainedRecompute", True)`.
- If fine-grained recomputes is on Tools->Dependency Graph will show the dependencies based on properties.
- Cyclic dependency checks are aware of input properties.
- The logic for fine-grained recomputes mimics the original logic as much as possible, to make minimal changes to the existing logic. This ensured that when the topological order is computed, we can still do that on the dependency graph based on objects.
- To make very clear where the logic has changed compared to coarse-grained recomputes, some dependency check codes has an extra level of indentation because of the condition if we are doing fine grained or coarse-grained recomputes. I think this benefits the review process where all new logic can be compared easily to the old logic. However, in terms of code-quality, I would prefer to refactor the code to get rid of this extra level of indentation.
- There is a test-suite that is explicit in whether a test should be run in either fine-grained, coarse-grained or both modes.
- In my humble opinion marking properties as "output" is misused in FreeCAD code to mean that a property change should not cause a recompute. For example `Label` and `Visibility` are marked as output properties. Although I would prefer those properties to change the flag to `NoRecompute`, it turns out that FreeCAD has semantics solidified in tests that prevent me to make this change. So, this PR is compatible with the semantics that output properties should not cause recomputes.

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->

Part of https://github.com/FreeCAD/FreeCAD/issues/12120
Closes #5948

<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->



<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

pieterhijma committed:
Core: Deprecate single precision for properties

pieterhijma committed:
Core: Introduce input properties

pieterhijma committed:
Core: Introduce dependency edges

pieterhijma committed:
Core: Compute in- and outlists based on properties

pieterhijma committed:
Core: Make cyclic dep checks aware of input props

pieterhijma committed:
Core: Create a fine-grained dependency graph

pieterhijma committed:
Core: Introduce logic to do fine-grained recompute

pieterhijma committed:
Gui: Allow setting properties as input

pieterhijma committed:
Core: Test fine-grained recomputes

eblanshey commented:
1. Do you know if this solution also addresses #5948?
2. Currently if a value such as "pad length" is edited in a way that results in the same value, it forces a recompute. E.g. if a raw value of 10mm is switched to an expression that results in 10mm, it recomputes. Does it make sense to make it "smarter" as part of this branch?

Thanks for your work, Pieter!

jdressu commented:
Kind of excited for this to hit! Amazing work everyone!
And i can't wait for phase 2. @pieterhijma Thank you so much!

I've been following your work on 'variant parts' for the past few months.
It was honestly a real turnoff for me, this past 4/5 years whenever i invested on 3d modeling with FreeCAD i eventually ended up stopping because of this. It kind of made no sense to me calling it a parametric without this feature.
I mean don't get me wrong, i know it is, and i appreciate all the work that's been done over the years, it's an amazing software, but it just didn't feel like "real" parametric design to me.

eblanshey commented:
I am also excited to see this work. What would happen if the spreadsheet or varset is in a separate file, and it is edited without the downstream parts opened? When I open the part file later, will it intelligently check if any of the dependent variables in external varsets/spreadsheets were updated? Just wondering about that. Thanks!

chennes commented on src/App/Application.h:
```diff
@@ -858,6 +858,8 @@ class AppExport Application
     /// Gets the base progress indicator instance.
     Base::ProgressIndicator& getProgressIndicator() { return _progressIndicator; }
 
+    bool fineGrained = true;
```
I don't love this as a public bool. This is a pretty fundamental change, I feel like this is a case where a getter/setter would make sense. Is this getting tied to a user preference someplace?

pieterhijma replied:
The ultimate plan is to replace the current recomputation mechanism with the fine-grained one. Before that I think it is wise to have it as an experimental feature first.

I removed the boolean altogether and the Python functions as well. I replaced it with a preference, so from the Preference editor -> General, you can turn on fine-grained recompute.

chennes reviewed (commented):
This is really nice, @pieterhijma -- I've noted some small things below, but overall it seems really good. I do wonder whether we really have to duplicate the two `_addBackLink/_removeBackLink` -> `_addBackLinkProp`/`_removeBackLinkProp` methods. There was no way to combine that? It feels like a maintenance issue (though no, I don't have a concrete solution for you, it's just a code smell).

jdressu commented:
Excitement levels back at zero. 🤣

I know open-source projects are hard to manage in a "wanting to follow the community's requests" sense..
But it saddens me when some of the really important changes don't get the attention they actually need.

(Don't want to get "down that rabbit hole".. But.. I'd argue if there was some kind of valid voting system in which freecad users could get a good picture of the weight of a functionality or fix this (the main task - [FEP-0010 Variant Parts](https://github.com/pieterhijma/FreeCAD-Enhancement-Proposals/blob/variant-parts/FEPs/FEP-0010-variant-parts/README.md) ) would most likely be there at the top.. but as it seems.. it is not.)

Was really hoping to see fase 2 somewhat in a 50%-80% complete state by now.. but..
i'm almost certain we will not be seeing that till mid of next year.. (Q2, Q3? probably? maybe? maybe not)

Anyway! Hope you all the best! Keep up the awesome work @pieterhijma !
I'll hopefully give freecad another go around 2 or 3 years from now in hopes of seeing "real parametric design" with those new 'Variant parts'.

Reqrefusion commented:
@pieterhijma What's the progress here? This PR will soon be entering its 6th month.

pieterhijma commented on src/App/FreeCADInit.py:
```diff
@@ -671,6 +671,7 @@ class PropertyType(IntEnum):
     Prop_Output = 8
     Prop_NoRecompute = 16
     Prop_NoPersist = 32
+    Prop_Input = 64
```
To change this feels like a very large responsibility. We have the following:

```
$ grep -R Prop_Input
App/FreeCADInit.py:    Prop_Input = 64
App/PropertyContainer.h:    Prop_Input   = 64,
App/PropertyContainerPyImp.cpp:    if (Type & Prop_Input) {
App/DocumentObject.cpp:    return (prop->getType() & Prop_Input) || prop->testStatus(Property::Input);
```

```
$ grep -R Property::Input
App/PropertyContainerPyImp.cpp:        statusMap["Input"] = Property::Input;
App/DocumentObject.cpp:    return (prop->getType() & Prop_Input) || prop->testStatus(Property::Input);
```

Here you can see that (from this PR by the way) "Input" is a string.

The output part is not from this PR and has been there for a long time:

<details>
<summary>grep of Prop_Output</summary>
<pre>
App/Annotation.cpp:    ADD_PROPERTY_TYPE(LabelText, (""), "Label", Prop_Output, "Text label of the annotation");
App/Annotation.cpp:    ADD_PROPERTY_TYPE(BasePosition, (Base::Vector3d()), "Label", Prop_Output, "Base position");
App/Annotation.cpp:    ADD_PROPERTY_TYPE(TextPosition, (Base::Vector3d()), "Label", Prop_Output, "Text position");
App/VRMLObject.cpp:                      static_cast<PropertyType>(Prop_ReadOnly | Prop_Output | Prop_Transient),
App/VRMLObject.cpp:                      static_cast<PropertyType>(Prop_ReadOnly | Prop_Output),
App/Property.cpp:    {App::Property::PropOutput,      Prop_Output},
App/FeatureTest.cpp:                      Prop_Output,
App/FeatureTest.cpp:                      (App::PropertyType)(Prop_Output | Prop_ReadOnly | Prop_Hidden),
App/FeatureTest.cpp:    ADD_PROPERTY_TYPE(Value, (0L), "Test", App::Prop_Output, "");
App/FeatureTest.cpp:    ADD_PROPERTY_TYPE(Value, (0L), "Test", App::Prop_Output, "");
App/FeatureTest.cpp:    ADD_PROPERTY_TYPE(Valid, (false), "Test", PropertyType(Prop_Output | Prop_ReadOnly), "");
App/FeatureTest.cpp:    ADD_PROPERTY_TYPE(MultLeft, (Base::Placement()), "Test", Prop_Output, "");
App/FeatureTest.cpp:    ADD_PROPERTY_TYPE(MultRight, (Base::Placement()), "Test", Prop_Output, "");
App/GeoFeature.cpp:                    (App::PropertyType)(Prop_Output | Prop_Hidden | Prop_Transient),
App/Link.cpp:                                PropertyType(Prop_Hidden | Prop_Output),
App/Link.cpp:                        || (prop.getType() & Prop_Output) || prop.testStatus(Property::Output)) {
App/FreeCADInit.py:    Prop_Output = 8
App/Property.h:        /// Corresponds to Prop_Output.
App/PropertyContainer.h:    Prop_Output      = 8,
App/PropertyContainerPyImp.cpp:    if (Type & Prop_Output) {
App/PropertyExpressionEngine.cpp:                    prop->testStatus(App::Property::Output) || (prop->getType() & App::Prop_Output);
App/Document.cpp:                                              Prop_Output | Prop_Hidden));
App/Document.cpp:                                              Prop_Output | Prop_Hidden));
App/DocumentObject.cpp:    ADD_PROPERTY_TYPE(Label, ("Unnamed"), "Base", Prop_Output, "User name of the object (UTF8)");
App/DocumentObject.cpp:    return (prop->getType() & Prop_Output) || prop->testStatus(Property::Output);
App/DocumentObject.cpp:        if (!testStatus(ObjectStatus::NoTouch) && !(prop->getType() & Prop_Output)
Gui/ViewProviderLink.cpp:    if (ext && !(prop->getType() & App::Prop_Output) && !prop->testStatus(App::Property::Output)) {
Mod/Fem/App/FemConstraintBearing.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintBearing.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintFluidBoundary.cpp:    ADD_PROPERTY_TYPE(Reversed,(0),"FluidBoundary",(App::PropertyType)(App::Prop_ReadOnly|App::Prop_Output),
Mod/Fem/App/FemConstraintFluidBoundary.cpp:    ADD_PROPERTY_TYPE(DirectionVector,(Base::Vector3d(0,0,1)),"FluidBoundary",App::PropertyType(App::Prop_ReadOnly|App::Prop_Output),
Mod/Fem/App/FemConstraintGear.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintPulley.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintPulley.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintPulley.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:    ADD_PROPERTY_TYPE(ForceX, (0.0), "ConstraintRigidBody", App::Prop_Output, "Applied force in X-direction");
Mod/Fem/App/FemConstraintRigidBody.cpp:    ADD_PROPERTY_TYPE(ForceY, (0.0), "ConstraintRigidBody", App::Prop_Output, "Applied force in Y-direction");
Mod/Fem/App/FemConstraintRigidBody.cpp:    ADD_PROPERTY_TYPE(ForceZ, (0.0), "ConstraintRigidBody", App::Prop_Output, "Applied force in Z-direction");
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintRigidBody.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintTransform.cpp:        App::Prop_Output,
Mod/Fem/App/FemConstraintTransform.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraintTransform.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraint.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemConstraint.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output | App::Prop_Hidden),
Mod/Fem/App/FemConstraint.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output | App::Prop_Hidden),
Mod/Fem/App/FemConstraintForce.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/FemSolverObject.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Fem/App/WorkerExtension.cpp:        App::PropertyType(App::Prop_Transient | App::Prop_Hidden | App::Prop_Output),
Mod/Fem/App/WorkerExtension.cpp:            App::Prop_Transient | App::Prop_Hidden | App::Prop_ReadOnly | App::Prop_Output
Mod/Inspection/Gui/ViewProviderInspection.cpp:        (App::PropertyType)(App::Prop_Output | App::Prop_Hidden),
Mod/Measure/App/MeasureLength.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureAngle.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureArea.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureBase.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output | App::Prop_NoRecompute),
Mod/Measure/App/MeasureDiameter.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureDistance.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasurePosition.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Measure/App/MeasureRadius.cpp:        App::PropertyType(App::Prop_ReadOnly | App::Prop_Output),
Mod/Mesh/App/MeshFeature.cpp:    ADD_PROPERTY_TYPE(Mesh, (MeshObject()), 0, App::Prop_Output, "The mesh kernel");
Mod/Part/App/FeaturePartBoolean.cpp:        (App::PropertyType)(App::Prop_Output | App::Prop_Transient | App::Prop_Hidden),
Mod/Part/App/FeaturePartCommon.cpp:        (App::PropertyType)(App::Prop_Output | App::Prop_Transient | App::Prop_Hidden),
Mod/Part/App/FeaturePartFuse.cpp:        (App::PropertyType)(App::Prop_Output | App::Prop_Transient | App::Prop_Hidden),
Mod/Part/App/PropertyTopoShape.cpp:        App::Prop_NoPersist | App::Prop_Output | App::Prop_Hidden
Mod/PartDesign/App/DatumLine.cpp:        App::Prop_Output,
Mod/PartDesign/App/DatumLine.cpp:    ADD_PROPERTY_TYPE(Length, (20), "Size", App::Prop_Output, "Length of the line");
Mod/PartDesign/App/DatumPlane.cpp:        App::Prop_Output,
Mod/PartDesign/App/DatumPlane.cpp:    ADD_PROPERTY_TYPE(Length, (20), "Size", App::Prop_Output, "Length of the plane");
Mod/PartDesign/App/DatumPlane.cpp:    ADD_PROPERTY_TYPE(Width, (20), "Size", App::Prop_Output, "Width of the plane");
Mod/PartDesign/App/Feature.cpp:        (App::PropertyType)(App::Prop_ReadOnly | App::Prop_Hidden | App::Prop_Output
Mod/PartDesign/App/ShapeBinder.cpp:    ADD_PROPERTY_TYPE(ClaimChildren, (false), "Base", App::Prop_Output, "Claim linked object as children");
Mod/Sketcher/App/SketchObject.cpp:                      (App::PropertyType)(App::Prop_Output | App::Prop_ReadOnly | App::Prop_Hidden),
Mod/Spreadsheet/App/Sheet.cpp:        (PropertyType)(Prop_ReadOnly | Prop_Hidden | Prop_Output),
Mod/Spreadsheet/App/Sheet.cpp:        (PropertyType)(Prop_ReadOnly | Prop_Hidden | Prop_Output),
Mod/TechDraw/App/DrawSVGTemplate.cpp:    ADD_PROPERTY_TYPE(PageResult, (nullptr),  group, App::Prop_Output,    "Embedded SVG code for template. For system use.");   //n/a for users
Mod/TechDraw/App/DrawViewDimExtent.cpp:    ADD_PROPERTY_TYPE(Source, (nullptr, nullptr), "", (App::PropertyType)(App::Prop_Output), "View containing the  dimension");
Mod/TechDraw/App/DrawViewDimExtent.cpp:    ADD_PROPERTY_TYPE(Source3d, (nullptr, nullptr), "", (App::PropertyType)(App::Prop_Output), "3D geometry to be dimensioned");
Mod/TechDraw/App/DrawViewDimExtent.cpp:    ADD_PROPERTY_TYPE(DirExtent ,(0), "", App::Prop_Output, "Horizontal / Vertical");
Mod/TechDraw/App/DrawViewDimExtent.cpp:    ADD_PROPERTY_TYPE(CosmeticTags ,(""), "", App::Prop_Output, "Id of cosmetic endpoints");
Mod/TechDraw/App/LandmarkDimension.cpp:    ADD_PROPERTY_TYPE(ReferenceTags, ("") , group, App::Prop_Output, "Tags of Dimension Endpoints");
Mod/TechDraw/App/CosmeticExtension.cpp:    EXTENSION_ADD_PROPERTY_TYPE(CosmeticVertexes, (nullptr), cgroup, App::Prop_Output, "CosmeticVertex Save/Restore");
Mod/TechDraw/App/CosmeticExtension.cpp:    EXTENSION_ADD_PROPERTY_TYPE(CosmeticEdges, (nullptr), cgroup, App::Prop_Output, "CosmeticEdge Save/Restore");
Mod/TechDraw/App/CosmeticExtension.cpp:    EXTENSION_ADD_PROPERTY_TYPE(CenterLines ,(nullptr), cgroup, App::Prop_Output, "CenterLine Save/Restore");
Mod/TechDraw/App/CosmeticExtension.cpp:    EXTENSION_ADD_PROPERTY_TYPE(GeomFormats ,(nullptr), cgroup, App::Prop_Output, "Geometry format Save/Restore");
Mod/TechDraw/App/DrawPage.cpp:                      (App::PropertyType)(App::Prop_Output), "Keep page in sync with model");
Mod/TechDraw/App/DrawView.cpp:    ADD_PROPERTY_TYPE(LockPosition, (false), group, App::Prop_Output, "Lock View position to parent Page or Group");
Mod/TechDraw/App/DrawView.cpp:    ADD_PROPERTY_TYPE(Rotation, (0.0), group, App::Prop_Output, "Rotation in degrees counterclockwise");
Mod/TechDraw/App/DrawView.cpp:    ADD_PROPERTY_TYPE(ScaleType, (prefScaleType()), group, App::Prop_Output, "Scale Type");
Mod/TechDraw/App/DrawView.cpp:    ADD_PROPERTY_TYPE(Caption, (""), group, App::Prop_Output, "Short text about the view");
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:    ADD_PROPERTY_TYPE(Arbitrary, (false), "Format", App::Prop_Output, "Value overridden by user");
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output, 
Mod/TechDraw/App/DrawViewDimension.cpp:                      App::Prop_Output,
Mod/TechDraw/App/DrawViewDimension.cpp:    ADD_PROPERTY_TYPE(LineAngle, (0.0), "Override", App::Prop_Output, "Dimension line angle");
Mod/TechDraw/App/DrawViewDimension.cpp:    ADD_PROPERTY_TYPE(ExtensionAngle, (0.0), "Override", App::Prop_Output, "Extension line angle");
Mod/TechDraw/App/DrawViewDimension.cpp:    ADD_PROPERTY_TYPE(UseActualArea, (true), "Area", App::Prop_Output,
Mod/TechDraw/App/DrawViewSection.cpp:                      App::Prop_Output,
Mod/Test/Document.py:    def testProp_OutputPropertyLink(self):
Mod/Test/Document.py:            FreeCAD.PropertyType.Prop_Output,
</pre>
</details>

<details>
<summary>grep of Poperty::Output</summary>
<pre>
App/DocumentObjectGroup.cpp:    _GroupTouched.setStatus(App::Property::Output, true);
App/GroupExtension.cpp:                    Property::Output,
App/GroupExtension.cpp:                // Temporary set the Property::Output on _GroupTouched, so that
App/Link.cpp:                    p->setStatus(Property::Output, true);
App/Link.cpp:                        || (prop.getType() & Prop_Output) || prop.testStatus(Property::Output)) {
App/PropertyContainerPyImp.cpp:        statusMap["Output"] = Property::Output;
App/PropertyExpressionEngine.cpp:                    prop->testStatus(App::Property::Output) || (prop->getType() & App::Prop_Output);
App/DocumentObject.cpp:    Label2.setStatus(App::Property::Output, true);
App/DocumentObject.cpp:    Visibility.setStatus(Property::Output, true);
App/DocumentObject.cpp:    return (prop->getType() & Prop_Output) || prop->testStatus(Property::Output);
App/DocumentObject.cpp:            && !prop->testStatus(Property::Output)) {
App/DocumentObject.cpp:    if (Visibility.testStatus(Property::Output)) {
Gui/Tree.cpp:    if (prop.testStatus(App::Property::Output) || prop.testStatus(App::Property::NoRecompute)) {
Gui/ViewProviderLink.cpp:    if (ext && !(prop->getType() & App::Prop_Output) && !prop->testStatus(App::Property::Output)) {
Mod/Fem/App/FemPostFilter.cpp:    XAxisData.setStatus(App::Property::Output, true);
Mod/Fem/App/FemPostFilter.cpp:    YAxisData.setStatus(App::Property::Output, true);
Mod/Fem/App/FemPostFilter.cpp:    PointData.setStatus(App::Property::Output, true);
Mod/Part/App/PreviewExtension.cpp:    PreviewShape.setStatus(App::Property::Output, true);
Mod/PartDesign/App/Body.cpp:    _GroupTouched.setStatus(App::Property::Output, true);
Mod/PartDesign/App/Body.cpp:    _GroupTouched.setStatus(App::Property::Output, true);
Mod/PartDesign/App/FeatureDressUp.cpp:    AddSubShape.setStatus(App::Property::Output, true);
Mod/PartDesign/App/ShapeBinder.cpp:                if (!prop.testStatus(App::Property::Output)
Mod/PartDesign/App/ShapeBinder.cpp:                                        prop->testStatus(App::Property::Output)
</pre>
</details>

Since setting a property to an output one can be done with a string in Python as well, I have the feeling that changing the name and leaving the strings "Output" and "Input" coexist will only add to the confusion.

<details>
<summary>grep of "Output"</summary>
<pre>
$ grep -R "\"Output\""
App/PropertyContainerPyImp.cpp:        ret.append(Py::String("Output"));
App/PropertyContainerPyImp.cpp:        statusMap["Output"] = Property::Output;
Gui/ReportView.cpp:    tabOutput->setWindowTitle(tr("Output"));
Gui/ReportView.cpp:        tabOutput->setWindowTitle(tr("Output"));
Mod/BIM/ArchCoveringGui.py:            if "ReadOnly" in status or "Output" in status:
Mod/CAM/Path/Main/Gui/Job.py:        if activate in ["Output", "Post Processor"]:
Mod/CAM/Path/Main/Job.py:    PostProcessorOutputFile = "Output"
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:            "Output",
Mod/CAM/Path/Main/Job.py:                "Output",
Mod/CAM/Path/Main/Job.py:                "Output",
Mod/CAM/Path/Main/Job.py:                "Output",
Mod/CAM/Path/Main/Job.py:                "Output",
Mod/CAM/Path/Main/Job.py:                "Output",
Mod/CAM/Path/Op/Helix.py:        obj.setPropertyStatus("Direction", ("ReadOnly", "Output"))
Mod/Fem/femsolver/calculix/calculixtools.py:            dat = self.obj.Document.addObject("App::TextDocument", self.obj.Name + "Output")
Mod/Fem/femsolver/elmer/elmertools.py:            dat = self.obj.Document.addObject("App::TextDocument", self.obj.Name + "Output")
Mod/Fem/femsolver/z88/z88tools.py:            dat = self.obj.Document.addObject("App::TextDocument", self.obj.Name + "Output")
Mod/PartDesign/App/DatumLine.cpp:    // their type is set to "Output" to avoid that they are marked as
Mod/PartDesign/App/DatumPlane.cpp:    // their type is set to "Output" to avoid that they are marked as
Mod/Robot/Gui/TrajectorySimulate.ui:     <widget class="QTextEdit" name="Output">
Mod/Spreadsheet/TestSpreadsheet.py:        ss2 = self.doc.addObject("Spreadsheet::Sheet", "Output")
Mod/Test/Recompute.py:        obj.addProperty("App::PropertyLength", "DoubleLength", "Output")
Mod/Test/Recompute.py:        obj.setPropertyStatus("DoubleLength", "Output")
Mod/Test/Recompute.py:        cube1.addProperty("App::PropertyLength", "OutputLength", "Output")
Mod/Test/Recompute.py:        cube1.setPropertyStatus("OutputLength", "Output")
</pre>
</details>

So, if the input/output part were only symbols, I wouldn't find it any problem to change it to `Prop_Driven` and `Prop_Driving`. But because Python can access these constructs through strings, I find it problematic, as external workbenches can rely on it as well.

What do you think?

chennes replied:
> But because Python can access these constructs through strings, I find it problematic, as external workbenches can rely on it as well.

This is true -- and indeed, two addons do use this, Assembly3 and Nodes. So a change from "Output" to "Driven" with no backwards-compatibility shim would have to be passed along to @realthunder and @j8sr0230. However, both of those Addons *only* use `setPropertyStatus` (not `getPropertyStatus`), so it would be simple to maintain a backwards-compatibility alias.

Then we really only have to worry about some potential future user of `getPropertyStatus` being confused when they set it to "Output" and then try to round-trip it and get "Driven" back. This seems like a pretty marginal thing to worry about, outside of just making sure it's documented.

I'd like to know @hyarion and @kadet1090's take on this. Is it better to just accept this somewhat odd usage of "Input" and "Output", or should we try to do better?

mnesarco replied:
There are also some Addons not under visibility of AddonManager index that may use the existing api.
i.e. https://github.com/mnesarco/fcapi/blob/b7c7f4acb5e48affc24d044db1a642f407b589e1/fpo.py#L471

But in this case I use int enums with the existing names, and some enum deserialization based on the name when required. (Ported directly from FreeCAD C++ source)

So changing existing Enum Symbols/Names/Strings may have a bigger impact than @realthunder and @j8sr0230 ...

hyarion replied:
> I'd like to know @hyarion and @kadet1090's take on this. Is it better to just accept this somewhat odd usage of "Input" and "Output", or should we try to do better?

While backwards compatibility is important, so is being able to improve our api if needed.
We've previously discussed ways to deprecate parts to allow doing modifications and it all comes down to communicate with our users that changes are happening and having a clear way on how to resolve issues due to the change.

We have a bunch of places where we handle api changes in dependencies by checking the library version so I don't think that requiring others to do the same when writing an addon for freecad is an extreme point of view, as long as it has a good reason behind the change and is well communicated.

maybe the mentioned compatibility shim for setPropertyStatus with warning in console + a note in release note could be enough?

FlachyJoe replied:
IMHO API deprecation is a good process in a dynamic project. We shouldn't keep misdone stuff to infinite future.
What could be done is a deprecation warning at (per session first) use for X releases before deletion. So the deprecated features should be listed somewhere or get clear comments we can search in the whole source tree.

pieterhijma replied:
I thought about it more: I'm not against deprecation, but then we should have a good way forward. Output properties gave quite some headaches while implementing this, see this comment (last bullet point from the PR description):

> - In my humble opinion marking properties as "output" is misused in FreeCAD code to mean that a property change should not cause a recompute. For example Label and Visibility are marked as output properties. Although I would prefer those properties to change the flag to NoRecompute, it turns out that FreeCAD has semantics solidified in tests that prevent me to make this change. So, this PR is compatible with the semantics that output properties should not cause recomputes.

Output properties are properties that can be written by objects and if they are written, they should of course not cause a recompute because the recompute would write the output properties again, giving an infinite loop. We could call them "Driven properties", but Visibility and Label are not really "driven". Properties that are not output properties and not input properties are currently more input/output properties. They can both be written and read during an execute. Often they are in the category "Driving", but they don't have to be. An example is Pad.Length that determines the height of a Pad, so a driving property. The constraint list of a Sketch is being written to and read so there it is already muddled what kind of property it should be.

The input properties that I define are more "pure input" properties. They can't be written to by the object. The original meaning was more that they are independent of the object in which they are defined. The most accurate term would perhaps be "independent property". For me the input/output scheme works well in my head. I consider objects as functions with in parameters (input properties), out parameters, output properties, and in/out parameters which are all the other properties.

I explain that independence in [this post](https://forum.freecad.org/viewtopic.php?p=795723#p795723) although I use different terms. For the sake of this discussion "exposed property" should be translated to "input property" and "You have to regard the property as not being part of it" should be read as the property is independent of the object. You should regard an input/exposed property as if it were defined in a separate document object.

In [this post](https://forum.freecad.org/viewtopic.php?p=795722#p795722) you can see the blue "Length" property that is independent of "Part".

chennes replied:
So the upshot is that basically there isn't a good name to use, because neither Input/Output nor Driving/Driven really capture the full essence of what we're doing here (and there is in fact a third "state" that is none of the above, or perhaps both of the above!). So... just leave it as it is here then?

pieterhijma replied:
Yes, that would be my recommendation. 

In terms of architecture, a first step would be to switch `Label`, `Visibility`, and others to `NoRecompute` instead of `Output` and deal with the problems that arise from that change. In the meantime, we may have marked more properties as `Input` by default, for example Cube length, width, and height. Then it becomes more clear to what extent the distinction is useful for FreeCAD, perhaps also for parallelism. At that stage, we may know what direction we want to go and we may have a chance to type these properties properly.

pieterhijma commented on src/App/PropertyLinks.cpp:
```diff
@@ -982,9 +989,11 @@ void PropertyLinkList::set1Value(int idx, DocumentObject* const& value)
             }
             if (value) {
                 value->_addBackLink(static_cast<DocumentObject*>(getContainer()));
+                obj->_addBackLinkProp(getName(), static_cast<DocumentObject*>(getContainer()));
```
I fixed it.

pieterhijma commented on src/App/Expression.cpp:
```diff
@@ -908,14 +908,15 @@ ExpressionDeps Expression::getDeps(int option)  const {
         const ObjectIdentifier &var = v.first;
         std::vector<std::string> strings;
-        for(auto &dep : var.getDep(false, &strings)) {
+        for(auto &dep : var.getDep(true, &strings)) {
```
I fixed it like what I wrote above.

pieterhijma commented on src/App/Document.cpp:
```diff
@@ -2418,6 +2418,28 @@ static void buildDependencyList(const std::vector<DocumentObject*>& objectArray,
+            }
+            // Add additional dependencies on specific properties unless it is
+            // an input property.  This to avoid over dependencies.
+            if (std::ranges::find(outList, objTo) == outList.end()
```
I made it linear.

pieterhijma commented:
The hold-up was the previous release. After that I had limited time and I waited a bit for the asynchronous recompute work as well as that could lead to problems in this work (it didn't lead to major problems).

pieterhijma commented:
> This is really nice, @pieterhijma -- I've noted some small things below, but overall it seems really good. I do wonder whether we really have to duplicate the two `_addBackLink/_removeBackLink` -> `_addBackLinkProp`/`_removeBackLinkProp` methods. There was no way to combine that? It feels like a maintenance issue (though no, I don't have a concrete solution for you, it's just a code smell).

I decided to basically duplicate the current interface to make clear what the difference exactly is. The main idea is that fine-grained recompute becomes the default. If that happens, the `_addBackLink(DocumentObject*)` calls can simply be removed.

This also makes it possible to run the two modes (fine-grained and coarse-grained recompute) side-by-side. Unchecking the preference of fine-grained recompute should give you FreeCAD exactly as before in terms of recomputation.

For now, it is a runtime switch. If the experience with fine-grained recompute is good, we can make it a compile-time switch. This duplication makes it possible to add ifdefs around it as well. Then, if the experience is good, a version later we can remove the coarse-grained code.

@chennes, could you please review again?

chennes commented on src/Mod/PartDesign/Gui/Command.cpp:
```diff
@@ -791,11 +792,24 @@ bool importExternalElements(App::PropertyLinkSub& prop, std::vector<App::SubObje
+        if (App::GetApplication().isFineGrainedRecomputeEnabled()) {
+            // Fully mimics the else block except for taking into account input properties.
+            for (const auto& [fromObj, fromProp, toObj, toProp] : inListProp) {
+                if (fromObj == sobj && !editObj->isInputProperty(toProp)) {
```
Might mean
```suggestion
                if (fromObj == sobj && !toObj->isInputProperty(toProp)) {
```
?

pieterhijma replied:
Definitely. This is incorrect. It turns out that this is not tested because it is Gui code. I will add a test for this as well.

chennes commented on src/App/DepEdge.h:
```diff
@@ -0,0 +1,124 @@
+#ifndef SRC_APP_DEPEDGE_H_
```
Missing license header.

pieterhijma replied:
Yep, resolved for all of them.

chennes commented on src/App/DepEdgePyImp.cpp:
```diff
@@ -0,0 +1,48 @@
+#include <App/DepEdge.h>
```
Missing license header.

chennes commented on src/App/DepEdge.pyi:
```diff
@@ -0,0 +1,17 @@
+from PyObjectBase import PyObjectBase
```
Missing license header.

chennes commented on src/App/Application.cpp:
```diff
@@ -819,6 +819,15 @@ bool Application::isAsyncRecomputeEnabled()
+        "User parameter:BaseApp/Preferences/General"
+    );
+    bool enableAsyncRecompute = hGrp->GetBool("FineGrainedRecompute");
+    return enableAsyncRecompute;
```
Copy/paste? Trivial, but you probably mean something like

```suggestion
    bool enableFineGrainedRecompute = hGrp->GetBool("FineGrainedRecompute");
    return enableFineGrainedRecompute;
```

chennes commented on src/App/DocumentObject.cpp:
```diff
@@ -259,6 +268,12 @@ bool DocumentObject::isTouched() const
     return ExpressionEngine.isTouched() || StatusBits.test(ObjectStatus::Touch);
 }
 
+void DocumentObject::enforceRecompute(std::string& propName)
```
Not a `const std::string &`?

pieterhijma replied:
Yes, an oversight, fixed.

chennes commented on src/App/Property.h:
```diff
@@ -95,8 +95,8 @@ class AppExport Property: public Base::Persistence
-        /// Whether a floating point number should be saved as single precision.
-        Single = 12,
+        /// Whether a property is an input property.
+        Input = 12,
```
I note that this will likely require an update to the `fcapi` library, @mnesarco .

mnesarco replied:
As far as all the breaking changes are documented, I can update fcapi and other things. But the important part is how to manage compatibility between versions. Addons need to be compatible with multiple FreeCAD versions over time while FreeCAD itself only cares about file compatibility. But that topic requires its own discussion.

chennes replied:
Fortunately I suspect no one was using `Single`, since it was apparently unimplemented.

chennes reviewed (commented):
Down to nitpicks.

pieterhijma committed:
Core: Apply suggestions from code review

Co-authored-by: Chris Hennes <chennes@pioneerlibrarysystem.org>

tritao commented on src/App/PropertyContainer.h:
```diff
@@ -70,6 +70,8 @@ enum PropertyType
     /// The property won't be saved to file at all.
     Prop_NoPersist   = 32,
+    /// The property is a true input property.
+    Prop_Input   = 64,
```
`Prop_Output` is referenced in a table in `Property.cpp`, should `Prop_Input` be there as well?

pieterhijma replied:
Yes indeed!

tritao commented on src/App/DepEdgePyImp.cpp:
```diff
@@ -0,0 +1,71 @@
+
+PyObject* DepEdgePy::getCustomAttributes(const char* /*attr*/) const
+{
+    return Py_None;
```
I think it is better to return `nullptr` here for unknown custom attributes.

pieterhijma replied:
Ok, great, will do.

tritao reviewed (commented):
Really great work!

The PR has many tests, but you mentioned that hidden references have their own dependency mechanism. Does it make sense to test whether fine-grained recompute handles the other dependency mechanism of hidden references still well?


Another minor nit, but I would consider renaming this property-level recompute as it seems a bit more descriptive, but since we're making this the default later it does not matter much.

pieterhijma commented:
> Really great work!

Thanks!

> The PR has many tests, but you mentioned that hidden references have their own dependency mechanism. Does it make sense to test whether fine-grained recompute handles the other dependency mechanism of hidden references still well?

Yes, indeed, it does make sense. I've added new tests and I found a small bug as well that is now fixed.
 
> Another minor nit, but I would consider renaming this property-level recompute as it seems a bit more descriptive, but since we're making this the default later it does not matter much.

I've always called it fine-grained recompute but property-level recompute could work as well. I don't think it makes much sense to change the PR name and the terms in the FEP but the user-facing preference can easily be renamed. If there are more people that have the preference "property-level recompute", I will change it.

Thanks for the review!

chennes reviewed (approved):
OK, I've re-reviewed and it looks like @pieterhijma has addressed all open concerns. I'm OK calling this "fine-grained recompute", and I think we can probably also choose to rename those GUI elements before a real release if weekly feedback indicates it's confusing. So my opinion is that this is ready to go in now, but I'll wait for @pieterhijma to make sure he doesn't have anything else he wants to do before that.

pieterhijma commented:
Ok, great. From my point of view, it is ready to go as well.

GeneGH commented:
This change makes the Dependency Graph close to useless for ordinary users trying to understand problems in a model file.
The complex view might be great for coders, but it destroys the simple view for users.

felixeriksson commented:
> This change makes the Dependency Graph close to useless for ordinary users trying to understand problems in a model file. The complex view might be great for coders, but it destroys the simple view for users.

Hi, do you have any example of the increased complexity that has arisen?

pieterhijma commented:
Hi, thanks for the information. The issue is already being discussed on [the forum](https://forum.freecad.org/viewtopic.php?t=106767) and in [this issue](https://github.com/FreeCAD/FreeCAD/issues/31727). See my [question on the forum](https://forum.freecad.org/viewtopic.php?p=900645#p900645). I'm interested in your opinions as well. Please let me know on the forum or in the issue.
